m3/reporter.go

```diff
@@ -243,6 +266,79 @@ func (r *reporter) AllocateTimer(
 	return cachedMetric{timer, r, size}
 }
 
+// AllocateHistogram implements tally.CachedStatsReporter.
+func (r *reporter) AllocateHistogram(
+	name string,
+	tags map[string]string,
+	buckets tally.Buckets,
+) tally.CachedHistogram {
+	var (
+		durations             = buckets.AsDurations()
+		lookupByValue         = buckets.AsValues()
+		lookupByDuration      = make([]int, buckets.Len())
+		cachedValueBuckets    []cachedMetric
+		cachedDurationBuckets []cachedMetric
+	)
+	bucketLenStr := strconv.Itoa(buckets.Len())
+	bucketLenStrLen := strconv.Itoa(len(bucketLenStr))
+	bucketIDFmt := "%0" + bucketLenStrLen + "d"
+	for i := 0; i < buckets.Len(); i++ {
+		lookupByDuration[i] = int(durations[i])
+	}
+	for i, pair := range tally.BucketPairs(buckets) {
+		valueTags, durationTags :=
+			make(map[string]string), make(map[string]string)
+		for k, v := range tags {
+			valueTags[k], durationTags[k] = v, v
+		}
+
+		idTagName := HistogramBucketIDTagName
+		nameTagName := HistogramBucketNameTagName
+
+		valueTags[idTagName] = fmt.Sprintf(bucketIDFmt, i)
+		valueTags[nameTagName] = fmt.Sprintf("%s-%s",
+			r.valueBucketString(pair.LowerBoundValue()),
+			r.valueBucketString(pair.UpperBoundValue()))
+
+		cachedValueBuckets = append(cachedValueBuckets,
+			r.allocateCounter(name, valueTags))
+
+		durationTags[idTagName] = fmt.Sprintf(bucketIDFmt, i)
```

Comment:
I can change this code to use the same bucket ID, they do have to use different names however.  Again this only happens at initial allocation time.